src/operator/roi_pooling.cc

```diff
@@ -0,0 +1,286 @@
+            assert(roi_batch_ind < batch_size_);
+            if (b != roi_batch_ind) {
+              continue;
+            }
+
+            int roi_start_w = round(bottom_rois[1] * spatial_scale_);
+            int roi_start_h = round(bottom_rois[2] * spatial_scale_);
+            int roi_end_w = round(bottom_rois[3] * spatial_scale_);
+            int roi_end_h = round(bottom_rois[4] * spatial_scale_);
+
+            bool in_roi = (w >= roi_start_w && w <= roi_end_w &&
+                           h >= roi_start_h && h <= roi_end_h);
+            if (!in_roi) {
+              continue;
+            }
+
+            // force malformed ROIs to be 1 * 1
+            int roi_height = max(roi_end_h - roi_start_h + 1, 1);
+            int roi_width = max(roi_end_w - roi_start_w + 1, 1);
+            const Dtype bin_size_h = static_cast<Dtype>(roi_height)
+                                     / static_cast<Dtype>(pooled_height_);
+            const Dtype bin_size_w = static_cast<Dtype>(roi_width)
+                                     / static_cast<Dtype>(pooled_width_);
+
+            // compute pooled regions correspond to original (h, w) point
+            int phstart = static_cast<int>(floor(static_cast<Dtype>(h - roi_start_h)
+                                                 * bin_size_h));
+            int pwstart = static_cast<int>(floor(static_cast<Dtype>(w - roi_start_w)
+                                                 * bin_size_w));
+            int phend = static_cast<int>(ceil(static_cast<Dtype>(h - roi_start_h + 1)
+                                              * bin_size_h));
+            int pwend = static_cast<int>(ceil(static_cast<Dtype>(w - roi_start_w + 1)
+                                              * bin_size_w));
+
+            // clip to boundaries of pooled region
+            phstart = min(max(phstart, 0), pooled_height_);
+            phend = min(max(phend, 0), pooled_height_);
+            pwstart = min(max(pwstart, 0), pooled_width_);
+            pwend = min(max(pwend, 0), pooled_width_);
+
+            // accumulate over gradients in pooled regions
+            int offset = (roi_n * channels_ + c) * pooled_height_ * pooled_width_;
+            const Dtype* offset_top_diff = top_diff + offset;
+            const Dtype* offset_argmax_data = argmax_data + offset;
+            for (int ph = phstart; ph < phend; ++ph) {
+              for (int pw = pwstart; pw < pwend; ++pw) {
+                const int pooled_index = ph * pooled_width_ + pw;
+                if (h * width_ + w == round(offset_argmax_data[pooled_index])) {
+                  gradient += offset_top_diff[pooled_index];
+                }
+              }
+            }
+
+            // Increment ROI data pointer
+            bottom_rois += bbox.size(2);
+          }
+          bottom_diff[offset_bottom_diff] = gradient;
+        }
+      }
+    }
+  }
+}
+}  // namespace mshadow
+
+namespace mxnet {
+namespace op {
+
+class ROIPoolingOp : public Operator {
```

Comment:
Defining two ROIPoolingOp will cause problems for amalgamation. Consider using template specialization to merge these two.